_data-prepper/common-use-cases/log-enrichment.md

```diff
@@ -0,0 +1,398 @@
+      {
+        "somekey" : "a",
+        "somevalue" : "val-a1",
+        "anothervalue" : "val-a2"
+      }  
+    ],
+    "b" : [
+      {
+        "somekey" : "b",
+        "somevalue" : "val-b1",
+        "anothervalue" : "val-b2"
+      },
+      {
+        "somekey" : "b",
+        "somevalue" : "val-b3",
+        "anothervalue" : "val-b4"
+      }
+    "c" : [
+      {
+        "somekey" : "c",
+        "somevalue" : "val-c1",
+        "anothervalue" : "val-c2"
+      }  
+    ]
+  }
+}
+```
+{% include copy-curl.html %}
+
+In many cases, you might want to flatten the array for each key. In these situations, you must choose only one object to remain. The processor offers the choice of either first or last.
+
+```json
+...
+processor:
+    - list_to_map:
+        key: "somekey"
+        source: "mylist"
+        target: "myobject"
+        flatten: true
+...
+```
+{% include copy-curl.html %}
+
+The incoming event structure is then flattened accordingly:
+
+```json
+{
+  "myobject" : {
+    "a" : {
+      "somekey" : "a",
+      "somevalue" : "val-a1",
+      "anothervalue" : "val-a2"
+    },
+    "b" : {
+      "somekey" : "b",
+      "somevalue" : "val-b1",
+      "anothervalue" : "val-b2"
+    }
+    "c" : {
+      "somekey" : "c",
+      "somevalue" : "val-c1",
+      "anothervalue" : "val-c2"
+    }
+  }
+}
+```
+{% include copy-curl.html %}
+
+## Processing incoming timestamps
+
+The [Date]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/date/) processor parses the timestamp key from incoming events by converting it to ISO 8601 format.
+
+```json
+...
+  processor:          
+    - date:
+        match:
+          - key: timestamp
+            patterns: ["dd/MMM/yyyy:HH:mm:ss"] 
+        destination: "@timestamp"
+        source_timezone: "America/Los_Angeles"
+        destination_timezone: "America/Chicago"
+        locale: "en_US"
+...
+```
+{% include copy-curl.html %}
+
+If the preceding pipeline processes the following event:
+
+```json
+{"timestamp": "10/Feb/2000:13:55:36"}
+```
+{% include copy-curl.html %}
+
+It converts the event into the following format:
+
+```json
+{
+  "timestamp":"10/Feb/2000:13:55:36",
+  "@timestamp":"2000-02-10T15:55:36.000-06:00"
+}
+```
+{% include copy-curl.html %}
+
+### Generating timestamps
+
+The Date processor can generate timestamps for incoming events if you specify `@timestamp` for the `destination` option.
+
+```json
+...
+    processor:
+    - date:
+        from_time_received: true
+        destination: "@timestamp"
+...
+```
+{% include copy-curl.html %}
+
+### Deriving punctuation patterns
+
+The [Substitute string]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/substitute-string/) processor (which is one of the Mutate string processors) lets you derive a punctuation pattern from incoming events. In the following example pipeline, the processor will scan incoming Apache log events and derive punctuation patterns from them.
```

Comment:
Can "Mutate string" be formatted so as to avoid any mid-sentence capitalization confusion?